### PR TITLE
Implement locale-specific bundle loading and `Stateful` registry.

### PR DESCRIPTION
**Pull request by mwistrand**
This PR includes the following i18n features:

* Locale-specific message loading
* `Stateful` registry
* Locale switching

Other features that will be added in the future pending discussions are message/date/time formatting and incorporation into the `dojo-app` ecosystem (I will create additional issues as needed).

@agubler had mentioned that it would be nice to have intellisense on i18n bundles. Given that these will be TypeScript modules, and since the bundle's default module will always be loaded for locale data, this could be accomplished by loading bundle defaults directly as normal modules instead of going through the loader (e.g., `import messages, { locales } from 'nls/main';`). However, we would then need a way to register that module with the `i18n` system so that it does not attempt to reload that package when the locale is switched or when `Stateful`s are registered. I am open to suggestions here.

### CODE REVIEW

**Comment by kitsonk** [comment]
Just to summerise a discussion that we had between @mwistrand @agubler and myself... looking at the API from a consumer's perspective, we would expect the following interfaces:

```typescript
interface LocaleContext {
    state: { locale?: string; };
}

interface I18N<T> {
    /* uses the ambient context to resolve the bundle */
    (bundle: T): Promise<T>;

    /* uses the context provided to resolve the bundle */
    (bundle: T, context: LocaleContext): Promise<T>;

    /* uses the specified locale to resolve the bundle */
    (bundle: T, locale: string): Promise<T>;
}
```

We also envisioned that from a consumer perspective, it would work something like this:

```typescript
import messages from './bundles/messages';
import i18n from 'dojo-i18n/i18n';

const context = {
    state: {
        locale: 'fr'
    }
};

/* Promise with ambient locale */
i18n(messages)
    .then((bundle) => {
        bundle.greeting; // 'Hello'
    });

/* await with contextual locale */
const bundleFr = await i18n(messages, context);
bundleFr.greeting; // 'Bonjour'

/* await with asserted locale */
const bundleDe = await i18n(messages, 'de');
bundleDe.greeting; // 'Hallo'
```

**Comment by mwistrand** [comment]
In order for the loader to be able to retrieve the correct locale bundle, will the `i18n` function also need a path (or will the bundle itself need to register its own path)?

**Comment by kitsonk** [comment]
I think between `i18n()` or whatever is in the bundle should resolve the path to the specific locale.

**Comment by mwistrand** [comment]
So the bundle passed into `i18n` would need to have something like a `bundlePath` property? By the time the object is passed into the `i18n` function, it's just an object without any sort of context for where its locale-specific messages reside. One option is to expose a registry that bundles add themselves to with a module ID that the loader can resolve to the correct locale bundle:

```
import registerBundle from 'dojo-i18n/i18n';

const messages = registerBundle('nls/main', {
  hello: 'Hello',
  helloReply: 'Hello',
  goodbye: 'Goodbye'
});

export default messages;
```

Another might be to require a path on the `i18n` function (imo, not the best solution):

```
import messages from './bundles/messages';
import i18n from 'dojo-i18n/i18n';

i18n(messages, 'app/bundles/messages', 'de');
```

**Comment by kitsonk** [comment]
While this isn't a fully fleshed out thought, why wouldn't a bundle define that path...  Something like this:

**interfaces.d.ts**

```typescript
export interface I18NLabels {
    [label: string]: string | () => string;
}

export interface I18NBundle<L extends I18NLabels> {
    baseUrl: string;
    labels: L;
}
```

**message.ts**

```typescript
import { I18NLabels, I18NBundle } from 'interfaces';

interface MessageLabels extends I18NLabels {
    greeting: string;
}

default export <I18NBundle<MessageLabels>> {
    baseUrl: './messages',
    labels: {
        greeting: 'Hello'
    }
};
```

**i18n.ts**

```typescript
import { I18NLabels, I18NBundle } from 'interfaces';

function i18n<T extends I18NLabels>(bundle: I18NBundle<T>): Promise<T>;
function i18n<T extends I18NLabels>(bundle: I18NBundle<T>, context: LocaleContext): Promise<T>;
function i18n<T extends I18NLabels>(bundle: I18NBundle<T>, locale: string): Promise<T>;
function i18n(bundle: I18NBundle<any>, context: LocaleContext | string): Promise<T> {
    /* implementation */
}
```

**Comment by mwistrand** [comment]
That works. I was stuck on the idea that the bundle would be a map of keys to messages, but this is ultimately better.

**Comment by kitsonk** [comment]
Yeah, so `i18n` extracts the shape, and then mixins in whatever bundle is resolved and returns that in the promise.  You will already have a default structure there in `bundle.labels` to mixin to!

**Comment by kitsonk** [comment]
It also helps if for some reason there is other meta data we have to end up supplying with the bundle...  The beauty of generics, once you spend a year mind-melding with them.

**Comment by mwistrand** [comment]
The basic loading functionality is consolidated within the `i18n` module, exposing an `i18n` function as discussed. Please also take a look at the README to ensure the example using [widgets](https://github.com/dojo/widgets) makes sense.

File: .travis.yml [hunk 1]
```diff
@@ -1,9 +1,23 @@
 sudo: false
 language: node_js
 node_js:
-  - "0.10"
+- '6'
+env:
+  global:
+  # Please get your own free key if you want to test yourself
+  - BROWSERSTACK_USERNAME: << USERNAME GOES HERE >>
```
Comment (kitsonk) [inline .travis.yml]: I would copy the keys over from another repo.

File: src/i18n.ts [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+/**
+ * A default bundle used as basis for loading locale-specific bundles.
+ */
+export interface Bundle<T extends Messages> {
+	/**
+	 * The absolute module ID used by the loader to resolve paths to locale-specific paths. This MUST follow the format
+	 * "{basePath}{separator}{filename}". For example, if the module ID for a bundle is "nls/common", the loader will
+	 * look for locale-specific bundles at "nls/{locale}/common".
+	 */
+	baseUrl: string;
```
Comment (kitsonk) [inline src/i18n.ts]: `readonly` as well?

File: src/i18n.ts [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+
+	/**
+	 * A list of supported locales. Any included locale MUST have an associated bundle.
+	 */
+	locales?: string[];
```
Comment (kitsonk) [inline src/i18n.ts]: Is there a good reason why this is optional?

File: src/i18n.ts [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+
+interface BundleMap<T extends Messages> {
+	[key: string]: T;
+}
+
+/**
+ * Represents an object from which a locale can be read. If a context object also has both `invalidate` and `own`
+ * methods, it will be updated (invalidated) when the locale is changed.
+ */
+export interface LocaleContext {
```
Comment (kitsonk) [inline src/i18n.ts]: Just a note, this is a great example of where `dojo/interfaces` will help.  There should be appropriate interfaces in there that would make this work.

File: src/i18n.ts [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+	/**
+	 * The state object from which the locale string is read.
+	 */
+	state: { locale?: string };
```
Comment (kitsonk) [inline src/i18n.ts]: Would it make sense to extract this as a generic.  Something like:

```typescript
export interface LocaleState {
    locale?: string;
}

export interface LocaleContext<S extends LocaleState> {
    state: S;
}
```

File: package.json [hunk 6]
```diff
@@ -1,35 +1,48 @@
...
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/dojo/i18n.git"
+  },
+  "scripts": {
+    "prepublish": "grunt peerDepInstall",
+    "test": "grunt test"
+  },
+  "peerDependencies": {
+    "@reactivex/rxjs": "5.0.0-beta.6",
```
Comment (kitsonk) [inline package.json]: What is dragging this in?  I don't specifically remember ES Observables being used.

**Comment by kitsonk** [comment]
I also enabled Travis on this repo... so any next push to this should cause the CI to kick off (which will also notify codecov.io)

File: src/i18n.ts (hunk 3 above)
Comment (mwistrand) [inline src/i18n.ts]: Projects might not be localized beyond the default when they are first created, in which case having an array of locales would not accomplish anything.

File: src/i18n.ts (hunk 5 above)
Comment (mwistrand) [inline src/i18n.ts]: That would be more consistent with how state interfaces are defined elsewhere in Dojo 2.

File: src/i18n.ts (hunk 2 above)
Comment (mwistrand) [inline src/i18n.ts]: Come to think of it, is there a good reason why _everything_ on this shouldn't be read-only?

Comment (kitsonk) [inline src/i18n.ts]: No, especially if you want to indicate to someone they can't modify the bundle.  I forgot to check, do you have `strictNullChecks` and `noImplicitThis`?  The only problem you can run into is that if you "build up" the creation of the object, but you can work around by casting to this interface on the function return.

File: tsconfig.json [hunk 7]
```diff
@@ -1,33 +1,21 @@
...
+		"strictNullChecks": true,
 		"outDir": "_build/",
+		"baseUrl": "node_modules",
 		"removeComments": false,
+		"sourceMap": true,
 		"target": "es5",
-		"experimentalDecorators": true,
-		"inlineSources": true,
-		"inlineSourceMap": true
+		"moduleResolution": "classic"
```
Comment (agubler) [inline tsconfig.json]: Can we use `node` for module resolution please?

File: src/i18n.ts [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+
+/**
+ * @private
+ * A type guard that determines whether a context object is stateful and can be invalidated.
+ *
+ * @param value
+ * The target value.
+ */
+function isContextObject(value: any): boolean {
+	return value.state && typeof value.own === 'function' && typeof value.invalidate === 'function';
```
Comment (agubler) [inline src/i18n.ts]: Might be worth wrapping this in a `Boolean(value.state && typeof value.own === 'function' && typeof value.invalidate === 'function')`

File: src/i18n.ts [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+ *
+ * @param context
+ * An optional context containing the locale information.
+ */
+function resolveLocale(context?: any): string {
+	if (context) {
+		if (typeof context === 'string') {
+			return context;
+		}
+		else if (context.state.locale) {
```
Comment (agubler) [inline src/i18n.ts]: does this assume `context.state` is always there?

File: src/i18n.ts [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+ * The new locale.
+ */
+export function switchLocale(locale: string): void {
+	const previous = rootLocale;
+	rootLocale = locale;
+
+	if (previous !== rootLocale) {
+		contextObjects.forEach(function (context: LocaleContext) {
+			if (!context.state.locale) {
+				(<any> context).invalidate();
```
Comment (agubler) [inline src/i18n.ts]: why do we need to cast this to `<any>`?

File: tslint.json [hunk 11]
```diff
@@ -1,51 +1,69 @@
...
 		"no-any": false,
 		"no-arg": true,
 		"no-bitwise": false,
-		"no-console": false,
 		"no-consecutive-blank-lines": true,
+		"no-console": false,
 		"no-construct": false,
+		"no-constructor-vars": true,
 		"no-debugger": true,
 		"no-duplicate-key": true,
```
Comment (agubler) [inline tslint.json]: FYI this will be removed in the next major bump of tslint, might as well remove it now

File: typings.json [hunk 12]
```diff
@@ -0,0 +1,25 @@
+{
+	"name": "dojo-i18n",
+	"globalDependencies": {
+		"extra-rxjs": "github:dojo/typings/custom/dojo2-extras/rxjs.d.ts#004fc065b3b24412c9b6575cdc06545964c370b5"
+	},
+	"peerDependencies": {
+		"@reactivex/RxJS": "npm:@reactivex/rxjs"
```
Comment (agubler) [inline typings.json]: can remove this and the `extra-rxjs` typings.

**Comment by agubler** [comment]
Generally I think throughout we could use function shorthand (arrows) across the changes, especially for anonymous functions like callbacks.

There is an on going discussion on this topic that can be found https://github.com/dojo/meta/issues/45#issuecomment-252639229

File: src/i18n.ts (hunk 9 above)
Comment (mwistrand) [inline src/i18n.ts]: A `state` object is required to meet the interface. Are there instances in which a `Stateful` object would not have a `state`?

File: src/i18n.ts (hunk 10 above)
Comment (mwistrand) [inline src/i18n.ts]: The `context` is only required to have a `state`; the `invalidate` and `own` methods are optional. The compiler was complaining about attempting to use this optional method, but I'll see if I can get this to work without casting.

Comment (kitsonk) [inline src/i18n.ts]: You should be able to deal with the strict null checks with `context.invalidate && context.invalidate();`

File: src/i18n.ts (hunk 4 above)
Comment (mwistrand) [inline src/i18n.ts]: I'm slightly confused what you mean. Are you saying that the entire `LocaleContext` interface should go into `dojo/interfaces`?

File: src/i18n.ts (hunk 2 above)
Comment (mwistrand) [inline src/i18n.ts]: Yes, both `strictNullChecks` and `noImplicitThis` are set  to `true` in `tsconfig.json`.

File: README.md [hunk 13]
```diff
@@ -1,10 +1,114 @@
...
+The examples below are provided in TypeScript syntax. The package does work under JavaScript, but for clarity, the examples will only include one syntax.
+
+### Message Bundle Loading
+
+`dojo-i18n` provides a means for loading locale-specific messages, and updating those messages when the locale changes. Each bundle has a default module that is `import`ed like any other TypeScript module. Locale-specific messages are then loaded via the `i18n` method. Every default bundle MUST provide a `baseUrl` that will be used to determine locale bundle locations, a `locales` array of supported locales, and a `messages` map of default messages. For example, suppose the module located at `nls/common.ts` contains the following contents:
+
+```typescript
+import { Bundle, Messages } from 'dojo-i18n/i18n';
+
+const baseUrl = 'nls/common';
```
Comment (agubler) [inline README.md]: is `baseUrl` the right name for this property, as it's not actually the `baseUrl`, for example if I had an nls file at `app/component/nls/widget.ts` I would assume the `baseUrl` would be `app/component/nls/` but it's not. According to the documentation it should be `app/component/nls/widget`... unless I completely misread this.

File: README.md [hunk 14]
```diff
@@ -1,10 +1,114 @@
...
+const locales = [ 'fr', 'ar', 'ar-JO' ];
+const messages = {
+	hello: 'Hello',
+	goodbye: 'Goodbye'
+};
+
+const bundle = { baseUrl, locales, messages };
+export default bundle as Bundle<Messages>;
```
Comment (mwistrand) [inline README.md]: Done. And as discussed, `baseUrl` is now `bundlePath`.

**Comment by agubler** [comment]
@mwistrand I think this is good to go, however just need to resolve the file conflicts (`.travis.yml`) and we can merged.

**Comment by kitsonk** [comment]
Why is there `i18n<Messages>` all over the place now?  Can't this be inferred contextually?

**Comment by agubler** [comment]
@kitsonk good point, I believe that they can be. As well as this there are lots of places that don't require explicit types (can be inferred from shape), unless there is a reason to keep them.

**Comment by agubler** [comment]
@mwistrand still needs conflicts fixing so it can be merged 😄

**Comment by mwistrand** [comment]
Git was not letting me rebase (`git add .; git rebase --continue` "there's nothing to add"), so I ended up having to cherry pick to make this work.

**Comment by agubler** [comment]
@mwistrand really strange that I could rebase from your branch 🎱

**Comment by agubler** [comment]
@mwistrand going to need to update ios to`9.3` in `intern-saucelabs.ts`, `7.1` it seems has been removed.

**Pull request merged**